### ISSUE TITLE
Question about ResourcePool

### ISSUE THREAD

**Issue by a-ucontrol** [ZLMediaKit/ZLMediaKit#3434]
My project for a very weak arm device where, ZLM is used as a library, has problems with memory allocation.

I saw ResourcePool in the code:

```c++
    template <typename C = FrameImp>
    static std::shared_ptr<C> create() {
#if 0
        static ResourcePool<C> packet_pool;
        static onceToken token([]() {
            packet_pool.setSize(1024);
        });
        auto ret = packet_pool.obtain2();
        ret->_buffer.clear();
        ret->_prefix_size = 0;
        ret->_dts = 0;
        ret->_pts = 0;
        return ret;
#else
        return std::shared_ptr<C>(new C());
#endif
    }
```

What is it for? How implemented and what are the results?

Thank you in advance

**Comment by suzp1984** [comment]
It seems a memory pool, an experimental feature(maybe), I think you can open this feature by `#if 1`.
https://github.com/ZLMediaKit/ZLMediaKit/blob/24689fefd1b913678a3aa4145c94a1f9fb68f2d6/src/Extension/Frame.h#L214-L234
Are you refer above code?
There are somewhere else has same macros, e.g. 
https://github.com/ZLMediaKit/ZLMediaKit/blob/24689fefd1b913678a3aa4145c94a1f9fb68f2d6/src/Rtsp/Rtsp.cpp#L623-L635

while some place used this `ResoucePool`. e.g. 
https://github.com/ZLMediaKit/ZLMediaKit/blob/24689fefd1b913678a3aa4145c94a1f9fb68f2d6/src/Record/MP4Demuxer.h#L75

The `ResourcePool` defined here: https://github.com/ZLMediaKit/ZLToolKit/blob/master/src/Util/ResourcePool.h

**Comment by xia-chu** [comment]
> has problems with memory allocation.

@a-ucontrol You mean there is a memory fragmentation issue?

**Comment by a-ucontrol** [comment]
> > has problems with memory allocation.
> 
> @a-ucontrol You mean there is a memory fragmentation issue?

The main problem is that the device only has 64 MB of memory. And, the system is heavily loaded during allocation and freeing.
Will ResourcePool help reduce the number of calls to the system to allocate memory?

**Comment by a-ucontrol** [comment]
@xia-chu I tried enabling ResourcePool for some parts of the code. Then I replaced jemalloc with tlsf, which allows you to allocate memory once, when the application starts. And it looks like there is no problem with memory usage in the C++ part of code.

As a result, I will try to abandon the use of webrtc in the simplest devices. But the question arises: how to display h264+opus on the web browser?

I create new issue with this question
